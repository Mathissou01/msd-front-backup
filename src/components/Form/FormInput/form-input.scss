@use "src/styles/01-settings/settings.colors" as colors;
@use "src/styles/02-tools/tools.inputs" as inputs;

.c-FormInput {
  display: flex;
  flex-direction: column;

  &_row {
    flex-direction: row;
  }

  textarea {
    min-height: 112px;
  }

  &__Label {
    margin-bottom: 2px;
    color: colors.$blue-2;
    font-weight: 700;
    font-size: 16px;
  }

  &__Input {
    @include inputs.input;

    &_invalid {
      @include inputs.input("invalid");
    }
  }
}
