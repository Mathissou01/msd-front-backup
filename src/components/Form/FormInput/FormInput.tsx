import _ from "lodash";
import classNames from "classnames";
import { ErrorMessage } from "@hookform/error-message";
import { ValidationRule } from "react-hook-form";
import { useFormContext } from "react-hook-form";
import React from "react";
import CommonErrorText from "../../Common/CommonErrorText/CommonErrorText";
import FormLabel, { LabelStyle, ValidationStyle } from "../FormLabel/FormLabel";
import "./form-input.scss";

interface IFormInputProps {
  type?: "number" | "text" | "email" | "password" | "hidden";
  name: string;
  label?: string;
  secondaryLabel?: string;
  validationLabel?: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  minLengthValidation?: number;
  maxLengthValidation?: number;
  patternValidation?: ValidationRule<RegExp>;
  patternValidationErrorMessage?: string;
  lengthHardValidation?: boolean;
  defaultValue?: string;
  placeholder?: string;
  flexStyle?: "column" | "row";
  labelStyle?: LabelStyle;
  validationStyle?: ValidationStyle;
  tagType?: "input" | "textarea";
}

export default function FormInput({
  type,
  name,
  label,
  secondaryLabel,
  validationLabel,
  isRequired = false,
  isDisabled = false,
  minLengthValidation,
  maxLengthValidation,
  patternValidation,
  patternValidationErrorMessage = "Format incorrect",
  lengthHardValidation = true,
  defaultValue,
  placeholder,
  flexStyle = "column",
  labelStyle,
  validationStyle = "inline",
  tagType = "input",
}: IFormInputProps) {
  /* Static Data */
  const errorMessages = {
    required: "Ce champ est obligatoire",
    minLength: `${minLengthValidation} caractères minimum`,
    maxLength: `${maxLengthValidation} caractères maximum`,
    pattern: patternValidationErrorMessage,
  };
  const Tag = tagType;

  /* Local Data */
  const {
    register,
    formState: { isSubmitting, errors },
  } = useFormContext();
  const inputClassNames = classNames("c-FormInput", {
    "c-FormInput_row": flexStyle === "row",
  });

  return (
    <div className={inputClassNames}>
      <FormLabel
        forId={name}
        label={label}
        isRequired={isRequired}
        secondaryLabel={secondaryLabel}
        validationLabel={validationLabel}
        flexStyle={flexStyle}
        labelStyle={labelStyle}
        validationStyle={validationStyle}
      >
        <Tag
          className={classNames("c-FormInput__Input", {
            "c-FormInput__Input_invalid": !!_.get(errors, name),
          })}
          {...register(name, {
            required: { value: isRequired, message: errorMessages.required },
            minLength:
              !lengthHardValidation && minLengthValidation
                ? {
                    value: minLengthValidation,
                    message: errorMessages.minLength,
                  }
                : undefined,
            maxLength:
              !lengthHardValidation && maxLengthValidation
                ? {
                    value: maxLengthValidation,
                    message: errorMessages.maxLength,
                  }
                : undefined,
            pattern: patternValidation ? patternValidation : undefined,
          })}
          type={type}
          id={name}
          defaultValue={defaultValue}
          minLength={lengthHardValidation ? minLengthValidation : undefined}
          maxLength={lengthHardValidation ? maxLengthValidation : undefined}
          placeholder={placeholder}
          disabled={isSubmitting || isDisabled}
          aria-invalid={!!_.get(errors, name)}
          aria-errormessage={`${name}_error`}
          data-testid="form-input"
        />
      </FormLabel>
      <ErrorMessage
        errors={errors}
        name={name}
        render={({ message }: { message: string }) => (
          <CommonErrorText message={message} errorId={`${name}_error`} />
        )}
      />
    </div>
  );
}
