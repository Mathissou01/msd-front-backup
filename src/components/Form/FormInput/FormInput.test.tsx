import { render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import React, { ReactNode } from "react";
import FormInput from "./FormInput";

const mock = {
  name: "form-input",
  label: "form input",
  secondaryLabel: "secondary label",
  validationLabel: "validation label",
};

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return (
    <>
      <FormProvider {...formMethods}>{props.children}</FormProvider>
    </>
  );
};

describe("FormInput", () => {
  it("renders", () => {
    const { container } = render(
      <Wrapper>
        <FormInput
          type={"text"}
          name={mock.name}
          label={mock.label}
          secondaryLabel={mock.secondaryLabel}
          validationLabel={mock.validationLabel}
          isDisabled={false}
        />
      </Wrapper>,
    );

    const input = screen.getByTestId("form-input");
    expect(input).toBeInTheDocument();
    expect(input).not.toHaveClass("c-FormInput__Input_invalid");
    expect(input).not.toHaveAttribute("disabled");

    expect(container).toMatchSnapshot();
  });

  it("can be disabled", () => {
    render(
      <Wrapper>
        <FormInput
          type={"text"}
          name={mock.name}
          label={mock.label}
          secondaryLabel={mock.secondaryLabel}
          validationLabel={mock.validationLabel}
          isDisabled
        />
      </Wrapper>,
    );

    const input = screen.getByTestId("form-input");
    expect(input).toHaveAttribute("disabled");
  });
});
