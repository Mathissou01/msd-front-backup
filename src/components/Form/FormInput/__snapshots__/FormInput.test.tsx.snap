// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput renders 1`] = `
<div>
  <div
    class="c-FormInput"
  >
    <div
      class="c-FormLabel"
    >
      <label
        class="c-FormLabel__LabelWrapper"
        for="form-input"
      >
        <span
          class="c-FormLabel__Label"
        >
          form input
        </span>
        <span
          class="c-FormLabel__Validation"
        >
          validation label
        </span>
      </label>
      <label
        class="c-FormLabel__LabelWrapper"
        for="form-input"
      >
        <span
          class="FormLabel__Secondary"
        >
          secondary label
        </span>
      </label>
      <div
        class="c-FormLabel__Content"
      >
        <input
          aria-errormessage="form-input_error"
          aria-invalid="false"
          class="c-FormInput__Input"
          data-testid="form-input"
          id="form-input"
          name="form-input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
